### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by JohnCoene** [BristolMyersSquibb/blockr#360] [General]
Roadmap for weeks and months ahead.

Feel free to give feedback.

## Short term

Pressing tangible features and issues that need to be addressed.

- Serialisation and restoration of stacks and workspace.
- Infinite reactive loop
- Stability: avoid crashes

### Infinite reactive loop

This is partly caused by the validation, though I believe @nbenn thought there were also other causes. Refactoring of the __validation__ is, as far as I can tell, required to fully fix the infinite reactive loop issue.

### Stability

I hope this is not too problematic to solve. We need to avoid app crashes at all costs. This mainly concerns block evaluation (`tryCatch`) but there may be other areas that need to be addressed.

## Medium to long-term

- Ability to selectively expose inputs
- Improved error messages
- Support table creation
- Library of stacks

We've received feedback that users would like to selectively expose inputs when a dashboard is locked: hiding all inputs is often not useful as users want to build interactive dashboards. Ideally users would be able to selectively choose which field remain visible when the dashboard is locked. Initial PR is https://github.com/blockr-org/blockr/pull/337 which is currently on draft. Feedback welcome.

Improved error messages: blockr may be more stable but will likely still occasionally throw errors; we can't control what users will write as blocks and these may errors. However, these errors are currently difficult for dashboard users to understand.

Support table creation. We've focused on plot outputs and we display a simple table by default, e.g.: on data block. But we are going to need to support table generation, primarily for life sciences where programmers are used to build tables by appending rows together.

Library of stacks. Oftentimes users create very similar stacks over and over for different projects, e.g.: demos that we have given already often repeat the creation of similar stacks. We'd need the possibility to create a library of stacks that dashboard builders could import. This is likely outside of blockr itself.

**Comment by JohnCoene** [comment]
@karmatarap @nbenn @christophsax @DivadNojnarg these are my thoughts for the roadmap, feel free to let me know what you think and what timelines you reckon is feasible.

**Comment by christophsax** [comment]
Thanks, @JohnCoene. We discussed with @karmatarap and internally and the division in short and medium-long term / backlog makes a lot of sense. We should focus on the short-term issues during May and try to finalize them as much as possible. Your short-term agenda, enriched with @nbenn's details, now looks like this:

Ideally, we can link all the tickets we work during May on to one of these. It also means other tasks / features should go into the backlog. 

# Short term (End of May)

## Save/restore

1. **Serialization**
    
    1. Full object serialization is functional
    2. Partial object serialization is for future consideration
    
2. **Deserialization**
    
    1. Full object deserialization might currently suffer from an ordering issue
        *   is currently being investigated
        *   we are lacking a reprex using the core UI

    
3. **versioning** (lower prio - perhaps after May?)
    
    1. No concrete plans here currently, might require "partial object serialization" to be of any practical use
    

## Stability issues/core fixes

1. **rework validation**, improve error messages displayed in UI
    
    1. Remove dependencies on `data` for block constructors
    2. Get rid of the dual constructor setup, i.e. only have `select_block()` instead of also having `new_select_block()`
    3. Same for field constructors -> constructor should not do initialization/validation for us
    4. Start removing "fixing things" in the validator functions
    5. Evaluation stops as soon as block validation fails
        *   eval stays in a try-catch
        *   An error here has similar consequences as a validation failure
    6. UI/communication component
    
2. **infinite reactivity**
    
    1. John looked into this and concluded that this is primarily caused by validation
    2. There might be other issues
    
3. **reduce app crashes**
    
    1. Field component functions have been wrapped with `try()` calls in `eval_set_field_values()`
    2. We should consider wrapping all block evaluations in a `tryCatch()`
    
4. extend testing infra
    
    1. David will look into this after his return from Japan

Here is the corresponding project board: https://github.com/orgs/blockr-org/projects/4. This roadmap is also available on the project board, if you click this button: 
<img width="283" alt="image" src="https://github.com/blockr-org/blockr/assets/1390827/c3fb76c9-b1c0-4f50-85c0-afc9a06662c3">

**Comment by christophsax** [comment]
### Blockr Summer 2024

Planning for our next monthly sprints:

Juni

*   topo-order awareness for fields and result blocks (Nicolas)
*   block server refactoring (David, discuss w/ Nicolas)
    *   remove block-level code output
*  add block UI

Juli

*   reconsider "recursive" fields vs field server approach (Nicolas)
    *   rethink list field
    *   case study mutate fields
*   core UI rework (David, discuss w/ John)
    *   shiny stuff in separate pkg?
    *   back to bslib for core UI

August

*   partial object serialization (Nicolas, maybe John?)
*   pretty printing of our objects (David?)

We can discuss tomorrow.

**Comment by christophsax** [comment]
## High-Level Goals

### End User Features

- Stack Gallery
- New Blocks: Summary Table Generation (Falcon), support for String Vector fields
- LLM Block

**Comment by JohnCoene** [comment]
Medium-long term

- Add blocks: hide blocks that are not compatible with current stack and/or display hint of block compatibility, e.g.: color border at bottom of block to match block pill color.
- Link stacks: it's still difficult to select a stack given we display the `id`, we should think of a solution.

**Comment by christophsax** [comment]
As discussed with Karma: markdownize a Stack is another user need.
